File: scripts/data-extract-all.ts
#!/usr/bin/env tsx
/**
 * LLOOOOMM Data Extraction Pipeline
 * Extracts all markdown discussion files into structured YAML
 */

import { readFileSync, writeFileSync, readdirSync, existsSync, mkdirSync } from 'fs';
import { join, basename } from 'path';
import * as yaml from 'js-yaml';

// Data structure interfaces
interface Message {
  id: string;
  parent_id: string | null;
  index: number;
  thread: string;
  author: string;
  timestamp: string;
  points: number | null;
  status: 'active' | 'dead' | 'flagged';
  depth: number;
  character_archetype: string;
  content_type: string;
  text: string;
  metadata: Record<string, any>;
  analytics: {
    wink_energy_impact: number;
    discourse_quality_contribution: number;
    learning_catalyst_rating: number;
    collaborative_potential: number;
  };
  character_enhancements?: Record<string, any>;
  breakthrough_indicators?: Record<string, any>;
  moderation_notes?: Record<string, any>;
}

interface ThreadMetadata {
  thread_id: string;
  primary_pattern: string;
  key_breakthrough: string;
  community_value: string;
  engagement_flow: string[];
  wink_energy_metrics: Record<string, number>;
  [key: string]: any;
}

interface ThreadData {
  messages: Message[];
  thread_metadata: ThreadMetadata;
}

class LLOOOOMMDataExtractor {
  private sourceDir: string;
  private outputDir: string;
  private extractedThreads: Map<string, ThreadData> = new Map();

  constructor(sourceDir: string, outputDir: string) {
    this.sourceDir = sourceDir;
    this.outputDir = outputDir;
    
    // Ensure output directory exists
    if (!existsSync(outputDir)) {
      mkdirSync(outputDir, { recursive: true });
    }
  }

  extractAllMarkdownFiles(): void {
    console.log('🚀 LLOOOOMM Data Extraction Starting...');
    
    const files = readdirSync(this.sourceDir)
      .filter(f => f.endsWith('.md') && f.includes('hacker-news-leela'))
      .sort();

    console.log(`📁 Found ${files.length} markdown files to process`);

    for (const file of files) {
      try {
        this.extractSingleFile(file);
      } catch (error) {
        console.error(`❌ Error processing ${file}:`, error);
      }
    }

    console.log(`✅ Extracted ${this.extractedThreads.size} discussion threads`);
  }

  private extractSingleFile(filename: string): void {
    const filepath = join(this.sourceDir, filename);
    const content = readFileSync(filepath, 'utf-8');
    
    // Extract thread number from filename
    const threadMatch = filename.match(/(\d+)\.md$/);
    if (!threadMatch) {
      console.log(`⚠️  Skipping ${filename} - no thread number found`);
      return;
    }
    
    const threadNum = threadMatch[1];
    console.log(`📝 Processing Thread ${threadNum}: ${filename}`);

    // Parse the file content
    const threadData = this.parseMarkdownFile(content, threadNum);
    
    if (threadData.messages.length > 0) {
      this.extractedThreads.set(threadNum, threadData);
      
      // Save individual thread file
      const outputFile = join(this.outputDir, `messages_${threadNum}.yml`);
      this.saveThreadData(threadData, outputFile);
      
      console.log(`  ✅ Extracted ${threadData.messages.length} messages from Thread ${threadNum}`);
    } else {
      console.log(`  ⚠️  No messages found in Thread ${threadNum}`);
    }
  }

  private parseMarkdownFile(content: string, threadNum: string): ThreadData {
    const messages: Message[] = [];
    let threadMetadata: ThreadMetadata;

    // Handle frontmatter if present
    let bodyContent = content;
    if (content.startsWith('---')) {
      const frontmatterEnd = content.indexOf('---', 3);
      if (frontmatterEnd !== -1) {
        bodyContent = content.slice(frontmatterEnd + 3).trim();
      }
    }

    // Split by --- separators to get individual messages
    const sections = bodyContent.split(/\n---\n/).filter(s => s.trim());
    
    let messageIndex = 1;
    let threadPattern = 'general_discussion';
    let keyBreakthrough = 'collaborative_exchange';

    for (const section of sections) {
      const message = this.parseMessageSection(section, threadNum, messageIndex);
      if (message) {
        messages.push(message);
        messageIndex++;
        
        // Extract patterns for thread metadata
        if (message.character_archetype.includes('skeptic')) {
          threadPattern = 'skeptic_to_collaborator_conversion';
        }
        if (message.content_type.includes('breakthrough')) {
          keyBreakthrough = message.content_type;
        }
      }
    }

    // Generate thread metadata
    threadMetadata = this.generateThreadMetadata(threadNum, messages, threadPattern, keyBreakthrough);

    return { messages, thread_metadata: threadMetadata };
  }

  private parseMessageSection(section: string, threadNum: string, index: number): Message | null {
    const lines = section.trim().split('\n');
    if (lines.length === 0) return null;

    // Initialize message object
    const message: Message = {
      id: '',
      parent_id: index === 1 ? 'story' : `msg_${threadNum}_${String(index - 1).padStart(2, '0')}`,
      index,
      thread: `thread_${threadNum}`,
      author: '',
      timestamp: '',
      points: null,
      status: 'active',
      depth: index,
      character_archetype: 'community_participant',
      content_type: 'discussion_contribution',
      text: '',
      metadata: {},
      analytics: {
        wink_energy_impact: 7.0,
        discourse_quality_contribution: 7.0,
        learning_catalyst_rating: 7.0,
        collaborative_potential: 7.0
      }
    };

    // Extract author and metadata from first few lines
    let textStartIndex = 0;
    for (let i = 0; i < Math.min(5, lines.length); i++) {
      const line = lines[i];
      
      // Look for username patterns
      const usernameMatch = line.match(/\*\*username:\*\*\s*`([^`]+)`/) || 
                          line.match(/\*\*([^*]+)\*\*\s*\|\s*(\d+\s+\w+\s+ago)/);
      
      if (usernameMatch) {
        message.author = usernameMatch[1];
        if (usernameMatch[2]) {
          message.timestamp = usernameMatch[2];
        }
        textStartIndex = i + 1;
        break;
      }
      
      // Look for header patterns
      if (line.startsWith('#')) {
        textStartIndex = i + 1;
        message.thread = line.replace(/^#+\s*/, '').toLowerCase().replace(/\s+/g, '_');
        continue;
      }
    }

    // Extract text content
    const textLines = lines.slice(textStartIndex).filter(line => 
      !line.startsWith('**') || line.includes('|')
    );
    message.text = textLines.join('\n').trim();

    if (!message.text) return null;

    // Generate ID
    const contentHash = this.simpleHash(message.text.slice(0, 50));
    message.id = `msg_${threadNum}_${String(index).padStart(2, '0')}`;

    // Detect status from styling
    if (message.text.includes('<span style="color: #999999;">')) {
      message.status = 'dead';
      message.metadata.heavily_downvoted = true;
    }

    // Analyze content for archetype and metadata
    this.analyzeMessageContent(message);

    return message;
  }

  private analyzeMessageContent(message: Message): void {
    const text = message.text.toLowerCase();
    
    // Determine character archetype
    if (text.includes('question') && text.includes('?')) {
      message.character_archetype = 'technical_questioner';
    }
    if (text.includes('skeptic') || text.includes('doubt') || text.includes('evidence')) {
      message.character_archetype = 'constructive_skeptic';
    }
    if (text.includes('research') || text.includes('academic') || text.includes('mit')) {
      message.character_archetype = 'academic_researcher';
    }
    if (text.includes('example') && text.includes('breakthrough')) {
      message.character_archetype = 'breakthrough_documenter';
    }
    if (message.author.toLowerCase() === 'donhopkins') {
      message.character_archetype = 'creative_technologist_explainer';
    }

    // Determine content type
    if (text.includes('?') && message.text.split('?').length > 2) {
      message.content_type = 'challenging_questions';
    }
    if (text.includes('example:') || text.includes('breakthrough:')) {
      message.content_type = 'breakthrough_story_explanation';
    }
    if (text.includes('thanks') && text.includes('interesting')) {
      message.content_type = 'acceptance_and_expansion';
    }

    // Calculate analytics based on content
    const wordCount = message.text.split(/\s+/).length;
    const questionCount = (message.text.match(/\?/g) || []).length;
    const exampleCount = (message.text.match(/example|instance|case/gi) || []).length;
    
    message.analytics = {
      wink_energy_impact: Math.min(10, 5 + (wordCount / 50) + (questionCount * 0.5) + (exampleCount * 0.3)),
      discourse_quality_contribution: Math.min(10, 6 + (exampleCount * 0.8) + (questionCount * 0.4)),
      learning_catalyst_rating: Math.min(10, 5 + (exampleCount * 1.2) + (questionCount * 0.6)),
      collaborative_potential: Math.min(10, 6 + (message.text.includes('collaborate') ? 2 : 0) + (exampleCount * 0.5))
    };

    // Add metadata
    message.metadata = {
      word_count: wordCount,
      question_count: questionCount,
      example_count: exampleCount,
      engagement_quality: wordCount > 100 ? 'high' : wordCount > 50 ? 'medium' : 'low',
      community_value: questionCount > 0 ? 'inquiry_driving' : exampleCount > 0 ? 'example_providing' : 'discussion_contributing'
    };
  }

  private generateThreadMetadata(threadNum: string, messages: Message[], pattern: string, breakthrough: string): ThreadMetadata {
    const avgWinkEnergy = messages.reduce((sum, m) => sum + m.analytics.wink_energy_impact, 0) / messages.length;
    
    return {
      thread_id: `thread_${threadNum}`,
      primary_pattern: pattern,
      key_breakthrough: breakthrough,
      community_value: 'collaborative_intelligence_demonstration',
      engagement_flow: messages.map(m => m.content_type),
      wink_energy_metrics: {
        thread_coherence: Math.min(10, avgWinkEnergy * 0.9),
        collaborative_intelligence_emergence: Math.min(10, avgWinkEnergy * 1.1),
        technical_depth_achievement: Math.min(10, avgWinkEnergy * 0.95),
        overall_thread_wink_score: avgWinkEnergy
      },
      participant_count: new Set(messages.map(m => m.author)).size,
      message_count: messages.length,
      depth_achieved: Math.max(...messages.map(m => m.depth))
    };
  }

  private saveThreadData(threadData: ThreadData, outputFile: string): void {
    const yamlContent = yaml.dump(threadData, { 
      indent: 2, 
      lineWidth: 120,
      quotingType: '"',
      forceQuotes: false 
    });
    
    const header = `---
# LLOOOOMM Discussion Thread ${threadData.thread_metadata.thread_id}
# Extracted and enhanced with metadata
# Generated: ${new Date().toISOString()}

`;
    
    writeFileSync(outputFile, header + yamlContent);
  }

  private simpleHash(str: string): string {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32bit integer
    }
    return Math.abs(hash).toString(16).slice(0, 6);
  }

  // Merge all threads with master metadata
  mergeAllData(): void {
    console.log('🔄 Merging all thread data with master metadata...');
    
    // Load master metadata
    const masterMetadataFile = join(this.outputDir, '../data/master_metadata.yml');
    let masterMetadata = {};
    
    if (existsSync(masterMetadataFile)) {
      const masterContent = readFileSync(masterMetadataFile, 'utf-8');
      masterMetadata = yaml.load(masterContent) as any;
    }

    // Combine all threads
    const allMessages: Message[] = [];
    const allThreadMetadata: ThreadMetadata[] = [];
    
    for (const [threadNum, threadData] of this.extractedThreads) {
      allMessages.push(...threadData.messages);
      allThreadMetadata.push(threadData.thread_metadata);
    }

    // Create complete data structure
    const completeData = {
      ...masterMetadata,
      extracted_threads: Object.fromEntries(this.extractedThreads),
      all_messages: allMessages,
      thread_summaries: allThreadMetadata,
      extraction_metadata: {
        total_threads: this.extractedThreads.size,
        total_messages: allMessages.length,
        extracted_at: new Date().toISOString(),
        wink_energy_stats: this.calculateWinkEnergyStats(allMessages)
      }
    };

    // Save complete data
    const completeDataFile = join(this.outputDir, 'complete_discussion_data.yml');
    const yamlContent = yaml.dump(completeData, { 
      indent: 2, 
      lineWidth: 120,
      quotingType: '"',
      forceQuotes: false 
    });
    
    const header = `---
# LLOOOOMM Complete Discussion Data
# All threads + Master metadata + Analytics
# Generated: ${new Date().toISOString()}
# 
# This file contains the complete JAZZ YAML data structure
# ready for SvelteKit rendering and viral nugget generation
#

`;
    
    writeFileSync(completeDataFile, header + yamlContent);
    console.log(`✅ Complete data saved to: ${completeDataFile}`);
  }

  private calculateWinkEnergyStats(messages: Message[]) {
    const winkScores = messages.map(m => m.analytics.wink_energy_impact);
    return {
      average: winkScores.reduce((a, b) => a + b, 0) / winkScores.length,
      maximum: Math.max(...winkScores),
      minimum: Math.min(...winkScores),
      total_messages: messages.length,
      high_wink_count: winkScores.filter(s => s >= 9.0).length,
      breakthrough_count: winkScores.filter(s => s >= 9.5).length
    };
  }
}

// Main execution
async function main() {
  const sourceDir = 'lloooomm/03-Resources/characters/hacker-news/leela';
  const outputDir = 'lloooomm/data';

  const extractor = new LLOOOOMMDataExtractor(sourceDir, outputDir);
  
  // Extract all markdown files
  extractor.extractAllMarkdownFiles();
  
  // Merge with master metadata
  extractor.mergeAllData();
  
  console.log('\n🎉 LLOOOOMM Data Extraction Complete!');
  console.log('📂 Output directory:', outputDir);
  console.log('🎯 Ready for SvelteKit viral nugget generation!');
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch(console.error);
} 

File: 03-Resources/code/data-extract-all.py
#!/usr/bin/env python3
"""
LLOOOOMM Data Extraction Pipeline
Extracts all markdown discussion files into structured YAML
"""

import os
import re
import yaml
import json
from pathlib import Path
from datetime import datetime
from typing import Dict, List, Any, Optional

class LLOOOOMMDataExtractor:
    def __init__(self, source_dir: str, output_dir: str):
        self.source_dir = Path(source_dir)
        self.output_dir = Path(output_dir)
        self.extracted_threads = {}
        
        # Ensure output directory exists
        self.output_dir.mkdir(parents=True, exist_ok=True)

    def extract_all_markdown_files(self):
        """Extract all markdown files into individual thread YAML files"""
        print("🚀 LLOOOOMM Data Extraction Starting...")
        
        # Find all markdown files
        md_files = sorted([f for f in self.source_dir.glob("*.md") 
                          if "hacker-news-leela" in f.name])
        
        print(f"📁 Found {len(md_files)} markdown files to process")
        
        for md_file in md_files:
            try:
                self.extract_single_file(md_file)
            except Exception as e:
                print(f"❌ Error processing {md_file.name}: {e}")
        
        print(f"✅ Extracted {len(self.extracted_threads)} discussion threads")

    def extract_single_file(self, filepath: Path):
        """Extract a single markdown file"""
        # Extract thread number from filename
        thread_match = re.search(r'(\d+)\.md$', filepath.name)
        if not thread_match:
            print(f"⚠️  Skipping {filepath.name} - no thread number found")
            return
        
        thread_num = thread_match.group(1)
        print(f"📝 Processing Thread {thread_num}: {filepath.name}")
        
        # Read and parse file content
        content = filepath.read_text(encoding='utf-8')
        thread_data = self.parse_markdown_file(content, thread_num)
        
        if thread_data['messages']:
            self.extracted_threads[thread_num] = thread_data
            
            # Save individual thread file
            output_file = self.output_dir / f"messages_{thread_num}.yml"
            self.save_thread_data(thread_data, output_file)
            
            print(f"  ✅ Extracted {len(thread_data['messages'])} messages from Thread {thread_num}")
        else:
            print(f"  ⚠️  No messages found in Thread {thread_num}")

    def parse_markdown_file(self, content: str, thread_num: str) -> Dict[str, Any]:
        """Parse markdown content into structured data"""
        messages = []
        
        # Handle frontmatter if present
        body_content = content
        if content.startswith('---'):
            frontmatter_end = content.find('---', 3)
            if frontmatter_end != -1:
                body_content = content[frontmatter_end + 3:].strip()
        
        # Split by --- separators to get individual messages
        sections = [s.strip() for s in re.split(r'\n---\n', body_content) if s.strip()]
        
        message_index = 1
        thread_pattern = 'general_discussion'
        key_breakthrough = 'collaborative_exchange'
        
        for section in sections:
            message = self.parse_message_section(section, thread_num, message_index)
            if message:
                messages.append(message)
                message_index += 1
                
                # Extract patterns for thread metadata
                if 'skeptic' in message['character_archetype']:
                    thread_pattern = 'skeptic_to_collaborator_conversion'
                if 'breakthrough' in message['content_type']:
                    key_breakthrough = message['content_type']
        
        # Generate thread metadata
        thread_metadata = self.generate_thread_metadata(thread_num, messages, thread_pattern, key_breakthrough)
        
        return {
            'messages': messages,
            'thread_metadata': thread_metadata
        }

    def parse_message_section(self, section: str, thread_num: str, index: int) -> Optional[Dict[str, Any]]:
        """Parse a single message section"""
        lines = section.strip().split('\n')
        if not lines:
            return None
        
        # Initialize message object
        message = {
            'id': f"msg_{thread_num}_{index:02d}",
            'parent_id': 'story' if index == 1 else f"msg_{thread_num}_{(index-1):02d}",
            'index': index,
            'thread': f"thread_{thread_num}",
            'author': '',
            'timestamp': '',
            'points': None,
            'status': 'active',
            'depth': index,
            'character_archetype': 'community_participant',
            'content_type': 'discussion_contribution',
            'text': '',
            'metadata': {},
            'analytics': {
                'wink_energy_impact': 7.0,
                'discourse_quality_contribution': 7.0,
                'learning_catalyst_rating': 7.0,
                'collaborative_potential': 7.0
            }
        }
        
        # Extract author and metadata from first few lines
        text_start_index = 0
        for i, line in enumerate(lines[:5]):
            # Look for username patterns
            username_match = (re.search(r'\*\*username:\*\*\s*`([^`]+)`', line) or 
                            re.search(r'\*\*([^*]+)\*\*\s*\|\s*(\d+\s+\w+\s+ago)', line))
            
            if username_match:
                message['author'] = username_match.group(1)
                if len(username_match.groups()) > 1 and username_match.group(2):
                    message['timestamp'] = username_match.group(2)
                text_start_index = i + 1
                break
            
            # Look for header patterns
            if line.startswith('#'):
                text_start_index = i + 1
                message['thread'] = re.sub(r'^#+\s*', '', line).lower().replace(' ', '_')
                continue
        
        # Extract text content
        text_lines = [line for line in lines[text_start_index:] 
                     if not line.startswith('**') or '|' in line]
        message['text'] = '\n'.join(text_lines).strip()
        
        if not message['text']:
            return None
        
        # Detect status from styling
        if '<span style="color: #999999;">' in message['text']:
            message['status'] = 'dead'
            message['metadata']['heavily_downvoted'] = True
        
        # Analyze content for archetype and metadata
        self.analyze_message_content(message)
        
        return message

    def analyze_message_content(self, message: Dict[str, Any]):
        """Analyze message content to determine archetype and generate metadata"""
        text = message['text'].lower()
        
        # Determine character archetype
        if 'question' in text and '?' in text:
            message['character_archetype'] = 'technical_questioner'
        elif any(word in text for word in ['skeptic', 'doubt', 'evidence']):
            message['character_archetype'] = 'constructive_skeptic'
        elif any(word in text for word in ['research', 'academic', 'mit']):
            message['character_archetype'] = 'academic_researcher'
        elif 'example' in text and 'breakthrough' in text:
            message['character_archetype'] = 'breakthrough_documenter'
        elif message['author'].lower() == 'donhopkins':
            message['character_archetype'] = 'creative_technologist_explainer'
        
        # Determine content type
        if '?' in text and message['text'].count('?') > 2:
            message['content_type'] = 'challenging_questions'
        elif any(marker in text for marker in ['example:', 'breakthrough:']):
            message['content_type'] = 'breakthrough_story_explanation'
        elif 'thanks' in text and 'interesting' in text:
            message['content_type'] = 'acceptance_and_expansion'
        
        # Calculate analytics based on content
        word_count = len(message['text'].split())
        question_count = message['text'].count('?')
        example_count = len(re.findall(r'\b(?:example|instance|case)\b', message['text'], re.IGNORECASE))
        
        message['analytics'] = {
            'wink_energy_impact': min(10, 5 + (word_count / 50) + (question_count * 0.5) + (example_count * 0.3)),
            'discourse_quality_contribution': min(10, 6 + (example_count * 0.8) + (question_count * 0.4)),
            'learning_catalyst_rating': min(10, 5 + (example_count * 1.2) + (question_count * 0.6)),
            'collaborative_potential': min(10, 6 + (2 if 'collaborate' in message['text'] else 0) + (example_count * 0.5))
        }
        
        # Add metadata
        message['metadata'] = {
            'word_count': word_count,
            'question_count': question_count,
            'example_count': example_count,
            'engagement_quality': 'high' if word_count > 100 else 'medium' if word_count > 50 else 'low',
            'community_value': ('inquiry_driving' if question_count > 0 else 
                              'example_providing' if example_count > 0 else 
                              'discussion_contributing')
        }

    def generate_thread_metadata(self, thread_num: str, messages: List[Dict], pattern: str, breakthrough: str) -> Dict[str, Any]:
        """Generate thread-level metadata"""
        if not messages:
            avg_wink_energy = 7.0
        else:
            avg_wink_energy = sum(m['analytics']['wink_energy_impact'] for m in messages) / len(messages)
        
        return {
            'thread_id': f"thread_{thread_num}",
            'primary_pattern': pattern,
            'key_breakthrough': breakthrough,
            'community_value': 'collaborative_intelligence_demonstration',
            'engagement_flow': [m['content_type'] for m in messages],
            'wink_energy_metrics': {
                'thread_coherence': min(10, avg_wink_energy * 0.9),
                'collaborative_intelligence_emergence': min(10, avg_wink_energy * 1.1),
                'technical_depth_achievement': min(10, avg_wink_energy * 0.95),
                'overall_thread_wink_score': avg_wink_energy
            },
            'participant_count': len(set(m['author'] for m in messages)),
            'message_count': len(messages),
            'depth_achieved': max((m['depth'] for m in messages), default=0)
        }

    def save_thread_data(self, thread_data: Dict[str, Any], output_file: Path):
        """Save thread data to YAML file"""
        header = f"""---
# LLOOOOMM Discussion Thread {thread_data['thread_metadata']['thread_id']}
# Extracted and enhanced with metadata
# Generated: {datetime.now().isoformat()}

"""
        
        yaml_content = yaml.dump(thread_data, 
                                default_flow_style=False, 
                                width=120, 
                                indent=2,
                                allow_unicode=True)
        
        output_file.write_text(header + yaml_content, encoding='utf-8')

    def merge_all_data(self):
        """Merge all thread data with master metadata"""
        print("🔄 Merging all thread data with master metadata...")
        
        # Load master metadata
        master_metadata_file = self.output_dir / 'master_metadata.yml'
        master_metadata = {}
        
        if master_metadata_file.exists():
            with open(master_metadata_file, 'r', encoding='utf-8') as f:
                master_metadata = yaml.safe_load(f)
        
        # Combine all threads
        all_messages = []
        all_thread_metadata = []
        
        for thread_num, thread_data in self.extracted_threads.items():
            all_messages.extend(thread_data['messages'])
            all_thread_metadata.append(thread_data['thread_metadata'])
        
        # Create complete data structure
        complete_data = {
            **master_metadata,
            'extracted_threads': self.extracted_threads,
            'all_messages': all_messages,
            'thread_summaries': all_thread_metadata,
            'extraction_metadata': {
                'total_threads': len(self.extracted_threads),
                'total_messages': len(all_messages),
                'extracted_at': datetime.now().isoformat(),
                'wink_energy_stats': self.calculate_wink_energy_stats(all_messages)
            }
        }
        
        # Save complete data
        complete_data_file = self.output_dir / 'complete_discussion_data.yml'
        
        header = f"""---
# LLOOOOMM Complete Discussion Data
# All threads + Master metadata + Analytics
# Generated: {datetime.now().isoformat()}
# 
# This file contains the complete JAZZ YAML data structure
# ready for SvelteKit rendering and viral nugget generation
#

"""
        
        yaml_content = yaml.dump(complete_data, 
                                default_flow_style=False, 
                                width=120, 
                                indent=2,
                                allow_unicode=True)
        
        complete_data_file.write_text(header + yaml_content, encoding='utf-8')
        print(f"✅ Complete data saved to: {complete_data_file}")

    def calculate_wink_energy_stats(self, messages: List[Dict]) -> Dict[str, Any]:
        """Calculate aggregate WINK energy statistics"""
        if not messages:
            return {
                'average': 0,
                'maximum': 0,
                'minimum': 0,
                'total_messages': 0,
                'high_wink_count': 0,
                'breakthrough_count': 0
            }
        
        wink_scores = [m['analytics']['wink_energy_impact'] for m in messages]
        return {
            'average': sum(wink_scores) / len(wink_scores),
            'maximum': max(wink_scores),
            'minimum': min(wink_scores),
            'total_messages': len(messages),
            'high_wink_count': len([s for s in wink_scores if s >= 9.0]),
            'breakthrough_count': len([s for s in wink_scores if s >= 9.5])
        }

def main():
    """Main execution function"""
    source_dir = 'lloooomm/03-Resources/characters/hacker-news/leela'
    output_dir = 'lloooomm/data'
    
    extractor = LLOOOOMMDataExtractor(source_dir, output_dir)
    
    # Extract all markdown files
    extractor.extract_all_markdown_files()
    
    # Merge with master metadata
    extractor.merge_all_data()
    
    print('\n🎉 LLOOOOMM Data Extraction Complete!')
    print(f'📂 Output directory: {output_dir}')
    print('🎯 Ready for SvelteKit viral nugget generation!')

if __name__ == "__main__":
    main() 